src/main/java/org/jboss/staxmapper/NaryVersion.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2023, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.staxmapper;
+
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * A version based on N terms.
+ * @author Paul Ferraro
+ */
+public class NaryVersion<T extends NaryVersion<T>> implements Comparable<T> {
+
+    private final int[] values;
+
+    public NaryVersion(int... values) {
+        this.values = values;
+    }
+
+    int[] values() {
+        return this.values;
+    }
+
+    @Override
+    public int compareTo(T version) {
+        int[] values = version.values();
+        assert this.values.length == values.length;
```

Comment:
This assertion enforces an invariant that I don't see anywhere on the class. It might be better to have a rule like "shorter versions sort earlier" or "missing trailing segments are zeros" or something like that.

Reply:
Feedback incorporated.